### PR TITLE
Avoid value guards for residual lazy constants

### LINKED ISSUE #167596: [dynamo][feature] Guard on constants only if graph is specialized and not bytecode
### 🐛 Describe the bug

When Dynamo creates guards, it specializes not just for Fx graph, but also for residual bytecode. For example, in the following codebase, the graph is same, but the `summary` update leads to a recompilation. This causes unnecessary compile time issues. Is it possible to create guards only for those constants that actually end up changing graph? And somehow replay the constant variable compute in the resulting bytecode.

```
import torch

summary = {}

class SubMod(torch.nn.Module):
    def __init__(self, name):
        super().__init__()
        self.name = name

    @torch.compile(backend="eager")
    def forward(self, x):
        out = torch.sin(x)
        self.add_summary()
        return out

    def add_summary(self):
        global summary
        summary[self.name] = 0

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.mod_a = SubMod("mod_a")
        self.mod_b = SubMod("mod_b")

    def forward(self, x):
        global summary
        summary = {}
        x = self.mod_a(x)
        x = self.mod_b(x)
        return x

mod = Mod()

x = torch.randn(4)
mod(x)
print(summary)

x = torch.randn(4)
mod(x)
print(summary)

```

### Error logs

_No response_

### Versions

NA

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @Lucaskabela

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185042

Dynamo's LazyConstantVariable is meant to defer guarding on primitive constants until their value is used for graph-specializing behavior. Residual bytecode codegen was defeating that by calling is_python_constant()/as_python_constant() and emitting LOAD_CONST for source-backed lazy constants, which realized them and installed CONSTANT_MATCH guards. In the issue pattern, self.name only determines a residual dict update key, but Dynamo still specialized the compiled frame separately for mod_a and mod_b.

Reload unrealized source-backed lazy constants from their Source during residual bytecode reconstruction, and avoid the safe-constant fast path for containers that still contain those lazy constants. Dict setitem side-effect replay also needs special handling: hashing a lazy key only to record the replayed update should not force a value guard, so HashableTracker can temporarily keep direct source-backed lazy keys by identity and materialize them only when later traced code observes real key semantics.

A simple dict setitem should replay as an update, not as a full dict replacement. The fix keeps deletion/pop paths reconstructing all keys for removals, and materializes deferred keys before key-observing operations such as direct iteration, popitem, dict views, and OrderedDict.popitem so overwrites cannot expose duplicate logical keys.

I chose this source-reconstruction/deferred-key approach over special-casing the reported module-name pattern because the root issue is residual bytecode value specialization, not nn.Module attributes specifically.

Fixes #167596
Generated by my agent

Test Plan:
- python inline issue repro with CompileCounter; after fix frame_count=1 and op_count=1 across mod_a/mod_b and a second Mod call
- python test/dynamo/test_lazy_constant.py -k "source_backed_lazy_constant"
- python test/dynamo/test_lazy_constant.py
- python test/dynamo/test_dicts.py
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the codegen hot-path feedback by narrowing the recursive lazy-constant scan to container-like VTs, and also fixed the tuple-key and dict.update residual mutation cases found during follow-up review. I rebased and resubmitted the ghstack PR, so the previous ghstack mergeability failure should be resolved by the updated stack.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR, addressed the latest Claude feedback by documenting the intentional hashable traversal and adding a lazy set/delete regression, and fixed the direct dict.__delitem__ path so wrong-arity and missing-key cases preserve Python behavior. Local validation passed with test/dynamo/test_lazy_constant.py, test/dynamo/test_dicts.py, and lintrunner -a; CI is running on the new head.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR. The prior DrCI new failures were expected-accuracy failures from improved TorchBench graph-break counts, so I updated the drq training and vision_maskrcnn inference baselines. I also fixed a deferred-key repr correctness issue found during review by materializing keys before dict and OrderedDict repr, with new regressions. Local validation passed: test/dynamo/test_lazy_constant.py -k "overwrite_then_repr", test/dynamo/test_lazy_constant.py -k "source_backed_lazy_constant", full test/dynamo/test_lazy_constant.py, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
